### PR TITLE
[#3] [Integrate] As a user, I can see a list of surveys on the home screen

### LINKED ISSUE #3: [Integrate] As a user, I can see a list of surveys on the home screen
## Why

Once logged in, users can navigate through a list of surveys that they can take. The mobile app will show the basic info of the survey as they scroll through each survey.

## Acceptance Criteria

- Use the survey cover image as the background image
    - Append `l` to the image URL to get the high-resolution image
- Use the survey title as the title label
- Use the survey description as the description label
- The number of page indicators must match the number of surveys

## Design

Screen Shot 2566-05-03 at 15 39 26

## Resources

[Home](https://www.figma.com/embed?embed_host=notion&url=https%3A%2F%2Fwww.figma.com%2Fproto%2FGjRPOjDyZ6f4EDL3wKarRK%2FChallenge-Mobile-App%3Fpage-id%3D0%253A1%26node-id%3D30%253A46%26scaling%3Dmin-zoom)

### PR DESCRIPTION
**Pull request by Wadeewee**
#3

## What happened 👀

- Use the survey cover image as the background image
    - Append `l` to the image URL to get the high-resolution image
- Use the survey title as the title label
- Use the survey description as the description label
- The number of page indicators must match the number of surveys

## Insight 📝

- Add [rxdart: ^0.27.7](https://pub.dev/packages/rxdart) Lib to use `BehaviorSubject stream` and `ZipStream` to merge the streams `profileStream` and `surveysStream` from calling APIs.
- Add more test cases for the `HomeViewModelTest`.

**Please note** the `Integration Test` for `HomeScreen` Will be added on this [story](https://github.com/users/Wadeewee/projects/1/views/1?pane=issue&itemId=28169013).

## Proof Of Work 📹

- When Fetch `profile & surveys` success:

https://github.com/Wadeewee/survey-flutter-ic-pooh/assets/28002315/f0063ce0-f9ef-4867-b957-c295acf3f05b

- When Fetch `profile` failed:

https://github.com/Wadeewee/survey-flutter-ic-pooh/assets/28002315/de2411d9-6336-441a-ba07-1997b86d60b5

- When Fetch `surveys` failed:

https://github.com/Wadeewee/survey-flutter-ic-pooh/assets/28002315/9324a353-fc4e-4bd4-b6f1-5f0c0a16193e

### CODE REVIEW

File: lib/ui/home/home_view_model.dart [hunk 1]
```diff
@@ -1,32 +1,102 @@
...
+  void loadData() async {
+    final profileStream = Stream.fromFuture(_getProfileUseCase.call());
+    final surveysStream = Stream.fromFuture(_getSurveysUseCase.call(
+      GetSurveysInput(
+        pageNumber: _defaultFirstPageIndex,
+        pageSize: _defaultPageSize,
+      ),
+    ));
+
+    ZipStream.zip2(
```
Comment (manh-t) [inline lib/ui/home/home_view_model.dart:70]: 👏

File: lib/ui/home/home_screen.dart [hunk 2]
```diff
@@ -103,38 +78,45 @@ class _HomeScreenState extends ConsumerState<HomeScreen> {
...
-              currentItem: _currentIndex,
-              count: mockSurveys.length,
-              alignment: Alignment.centerLeft,
-              padding: const EdgeInsets.symmetric(horizontal: space15),
-              margin: const EdgeInsets.symmetric(horizontal: space5),
-              size: const Size(
-                pagerIndicatorSize,
-                pagerIndicatorSize,
+          if (surveys.isNotEmpty)
+            Positioned(
```
Comment (hoangnguyen92dn) [inline lib/ui/home/home_screen.dart]: OPTIONAL: We could extract to new `_build` method to build the indicator widget.

File: lib/ui/home/home_view_model.dart [hunk 3]
```diff
@@ -1,32 +1,102 @@
...
+  final BehaviorSubject<String> _profileAvatar = BehaviorSubject();
+
+  Stream<String> get profileAvatar => _profileAvatar.stream;
+
+  final BehaviorSubject<List<SurveyModel>> _surveys = BehaviorSubject();
+
+  Stream<List<SurveyModel>> get surveys => _surveys.stream;
+
+  void loadData() async {
+    final profileStream = Stream.fromFuture(_getProfileUseCase.call());
```
Comment (hoangnguyen92dn) [inline lib/ui/home/home_view_model.dart]: OPTIONAL: We could use `.asStream()` to convert Future to Stream for usage

```suggestion
    final profileStream = _getProfileUseCase.call().asStream();
```

File: lib/ui/home/home_view_model.dart [hunk 4]
```diff
@@ -1,32 +1,102 @@
...
+      profileStream,
+      surveysStream,
+      (profileResult, surveysResult) {
+        _today.add(DateTime.now().getFormattedString());
+
+        if (profileResult is Success<ProfileModel>) {
+          _profileAvatar.add(profileResult.value.avatarUrl);
+        } else {
+          _profileAvatar.add('');
+        }
```
Comment (hoangnguyen92dn) [inline lib/ui/home/home_view_model.dart]: Should we skip emitting empty event? 🤔

Comment (Wadeewee) [reply lib/ui/home/home_view_model.dart]: As we ignore the error when fetch `profile` failed, I think we can @hoangnguyen92dn 👍

File: lib/ui/home/home_view_model.dart (hunk 3 above)
Comment (Wadeewee) [reply lib/ui/home/home_view_model.dart]: Good point 👍 @hoangnguyen92dn, Thanks for reminding me 🙏

**Review by manh-t — COMMENTED** [review]
only Huey's comments, rest LGTM 💯

**Review by hoangnguyen92dn — APPROVED** [review]

**Review by manh-t — APPROVED** [review]

**Pull request merged**